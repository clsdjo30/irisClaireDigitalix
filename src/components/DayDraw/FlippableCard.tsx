// FlippableCard.tsx
import React, { useEffect} from 'react';
import Animated, { useAnimatedStyle, useSharedValue, withTiming, interpolate } from 'react-native-reanimated';
import Card from '../Card';
import { ImageSourcePropType, ViewStyle } from 'react-native';

type FlippableCardProps = {
    source: ImageSourcePropType,
    onPress: () => void,
    flipped: boolean,
    isBackCard: boolean,
    style?: ViewStyle | ViewStyle[]
};

const FlippableCard: React.FC<FlippableCardProps> = ({ source, onPress, flipped, isBackCard, style }) => {
    const rotate = useSharedValue(flipped ? 1 : 0);

    useEffect(() => {
        rotate.value = flipped ? 1 : 0;
    }, [flipped]);

    const animatedStyles = useAnimatedStyle(() => {
        const rotateValue = interpolate(rotate.value, [0, 1], isBackCard ? [180, 360] : [0, 180]);
        return {
            transform: [
                { rotateY: withTiming(`${rotateValue}deg`, { duration: 1000 }) }
            ]
        };
    });

    return (
        <Animated.View  style={[animatedStyles, style]}>
            <Card onPress={onPress} source={source} />
        </Animated.View>
    );
};

export default FlippableCard;
