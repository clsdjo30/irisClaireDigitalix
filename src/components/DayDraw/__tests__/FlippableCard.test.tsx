import { useEffect, useRef, useState } from 'react';
import { renderHook, act } from '@testing-library/react-hooks';
import { render, fireEvent } from '@testing-library/react-native';
import { ImageSourcePropType } from 'react-native';
import FlippableCard from '../FlippableCard';

const useMyHook = () => {
  const rotate = useRef(0);
  const [flipped, setFlipped] = useState(false);

  // Code to be tested
  useEffect(() => {
    rotate.current = flipped ? 1 : 0;
  }, [flipped]);

  return { rotate, flipped, setFlipped };
};

describe('FlippableCard', () => {
  const mockProps = {
    source: require('../../../../assets/images/cards/back/Card_Back_001.png') as ImageSourcePropType,
    onPress: jest.fn(),
    flipped: false,
    isBackCard: false
  };

  it('renders correctly', () => {
    const tree = render(<FlippableCard {...mockProps} />);
    expect(tree).toMatchSnapshot();
  });

  

  it('calls onPress when the card is clicked', () => {
    const { getByTestId } = render(<FlippableCard {...mockProps} />);
    const card = getByTestId('card');

    fireEvent.press(card);

    expect(mockProps.onPress).toHaveBeenCalled();
  });
 
  
  it('should flip when pressed', () => {
    const { getByTestId } = render(
      <FlippableCard {...mockProps} />
    );
    const card = getByTestId('card');
    fireEvent.press(card);
    expect(card).toBeTruthy();
  });

  it('should not flip when pressed', () => {
    const { getByTestId } = render(
      <FlippableCard {...mockProps} />
    );
    const card = getByTestId('card');
    fireEvent.press(card);
    expect(card).toBeTruthy();
  });

  it('should update rotate value on flipped state change', () => {
    const { result } = renderHook(() => useMyHook());

    // Initial state
    expect(result.current.rotate.current).toEqual(0);
    expect(result.current.flipped).toEqual(false);

    act(() => {
      result.current.setFlipped(true);
    });

    // Dispose it since the ref object will hold the old value.
    renderHook(() => useMyHook()).unmount();

    // After flipping
    expect(result.current.rotate.current).toEqual(1);
    expect(result.current.flipped).toEqual(true);
  });

});
