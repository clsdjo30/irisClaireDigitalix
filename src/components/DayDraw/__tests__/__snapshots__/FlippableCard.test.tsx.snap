// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FlippableCard renders correctly 1`] = `
<View
  animatedStyle={
    {
      "value": {
        "transform": [
          {
            "rotateY": "0deg",
          },
        ],
      },
    }
  }
  collapsable={false}
  style={
    {
      "transform": [
        {
          "rotateY": "0deg",
        },
      ],
    }
  }
>
  <View
    accessibilityState={
      {
        "busy": undefined,
        "checked": undefined,
        "disabled": undefined,
        "expanded": undefined,
        "selected": undefined,
      }
    }
    accessibilityValue={
      {
        "max": undefined,
        "min": undefined,
        "now": undefined,
        "text": undefined,
      }
    }
    accessible={true}
    collapsable={false}
    focusable={true}
    onBlur={[Function]}
    onClick={[Function]}
    onFocus={[Function]}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      {
        "elevation": 5,
        "shadowColor": "#fff",
      }
    }
    testID="card"
  >
    <Image
      source={1}
      style={
        {
          "height": 150,
          "resizeMode": "cover",
          "width": 80,
        }
      }
      testID="card-image"
    />
  </View>
</View>
`;
